coremltools/converters/mil/frontend/torch/test/test_torch_ops.py

```diff
@@ -2785,56 +2785,85 @@ def forward(self, x):
         inputs = [torch.tensor([start, end, step])]
         self.run_compare_torch(inputs, model, backend=backend, input_as_shape=False)
 
-
 class TestEinsum(TorchBaseTest):
     @pytest.mark.parametrize(
-        "backend, equation, reverse_input_order",
+        "backend, equation, reverse_input_order, dynamic",
         itertools.product(
             backends,
-            ["abcd,adce->abce",
-             "abc,cbd->abd",
-             "bnqd,bnkd->bnqk",
-             "abc,cd->abd",
-             "abc,cde->abde",
-             "btnh,bfnh->bnft",
-             "bnft,btnh->bfnh",
-             "abcd,cde->abe",
-             "a b c d , a d c e -> a b c e"],
-            [False, True]
+            [
+                # Hardcoded cases
+                "abcd,adce->abce",
+                "abc,cbd->abd",
+                "bnqd,bnkd->bnqk",
+                "abc,cd->abd",
+                "abc,cde->abde",
+                "btnh,bfnh->bnft",
+                "bnft,btnh->bfnh",
+                "abcd,cde->abe",
+                "a b c d , a d c e -> a b c e",
+                # Generic cases
+                "i,i->i",
+                "i,j->ij",
+                "ab,b->a",
+                "ab,ab->b",
+                "abc,abc->a",
+                "abc,abc->c",
+                "abc,bac->c",
+                "abc,acd->abd",
+                "abc,abc->ab",
+                "abc,abc->bc",
+                "abc,bac->ba",
+                "abcd,cb->dca",
+                "abcd,acdb->ad",
+                "abcd,abde->abce",
+                "abcd,efbd->eafc",
+                "acdb,bade->abce",
+             ],
+            [False, True],
+            [False, True],
         ),
     )
-    def test_einsum(self, backend, equation, reverse_input_order):
+    def test_einsum(self, backend, equation, reverse_input_order, dynamic):
         class TestEinsum(nn.Module):
             def forward(self, x, y):
                 return torch.einsum(equation, x, y)
 
-        if equation in ["abcd,adce->abce", "a b c d , a d c e -> a b c e"]:
-            input_shapes = [[3, 4, 2, 6], [3, 6, 2, 2]]
-        elif equation == "abc,cbd->abd":
-            input_shapes = [[4, 2, 6], [6, 2, 2]]
-        elif equation == "bnqd,bnkd->bnqk":
-            input_shapes = [[1,2,3,4], [1,2,4,4]]
-        elif equation == "abc,cd->abd":
-            input_shapes = [[2,3,4], [4,5]]
-        elif equation == "abc,cde->abde":
-            input_shapes = [[2,3,4], [4,5,6]]
-        elif equation == "btnh,bfnh->bnft":
-            input_shapes = [[1,2,3,4], [1,5,3,4]]
-        elif equation == "bnft,btnh->bfnh":
-            input_shapes = [[1,2,3,4], [1,4,2,6]]
-        elif equation == "abcd,cde->abe":
-            input_shapes = [[1,2,3,4], [3,4,6]]
-        else:
-            raise ValueError("unrecognized equation")
+        def make_input_types(equation, dynamic):
```

Comment:
Actually, @fukatani  can you re0use this utility function for both tensorflow / torch tests?
Like put this function in a common test ultil file.
For the tensorflow, `dynamic` is always `False`,
 and only the first return value (`[a_shape, b_shape]`) is used.
That would remove the duplicate code.